test/controllers/csvs_controller_test.rb

```diff
@@ -0,0 +1,335 @@
+      end
+    end
+
+    describe "a GET to show for pending" do
+      setup { get :show, id: export.id, format: format}
+      
+      describe "as html" do
+        let(:format) { :html }
+      
+        it "renders the status template" do
+          assert_template :show
+          @response.body.must_include "being prepared"
+        end
+      end
+      
+      describe "as JSON" do
+        let(:format) { :json }
+        
+        it "renders json" do
+          assert_equal "application/json", @response.content_type
+          assert_response :ok
+          @response.body.must_equal export.to_json
+        end
+      end
+    end
+
+    describe "a GET to show for deleted" do
+      setup do
+        @csv_export = CsvExport.find(export.id)
+        @csv_export.deleted!
+        get :show, id: @csv_export.id, format: format
+      end
+      
+      describe "as html" do
+        let(:format) { :html }
+      
+        it "renders the status template" do
+          assert_template :show
+          @response.body.must_include "deleted"
+        end
+      end
+      
+      describe "as JSON" do
+        let(:format) { :json }
+        
+        it "renders json" do
+          assert_equal "application/json", @response.content_type
+          assert_response :ok
+          @csv_export.reload
+          @response.body.must_equal @csv_export.to_json
+        end
+      end
+    end
+
+    describe "a GET to show for invalid" do
+      setup { get :show, id: -1, format: format }
+
+      describe "as html" do
+        let(:format) { :html }
+
+        it "redirects to index with error flash" do
+          assert_redirected_to csvs_path
+          flash[:error].must_equal "The CSV export does not exist."
+        end
+      end
+
+      describe "as JSON" do
+        let(:format) { :json }
+
+        it "renders json" do
+          assert_equal "application/json", @response.content_type
+          assert_response 404
+          @response.body.must_equal "{\"status\":\"not-found\"}"
+        end
+      end
+    end
+
+    describe "a GET to show for finished" do
+      setup do
+        @csv_export = CsvExport.find(export.id)
+        @csv_export.finished!
+        get :show, id: @csv_export.id, format: format
+      end
+
+      describe "as html" do
+        let(:format) { :html }
+
+        it "renders the status template" do
+          assert_template :show
+          @response.body.must_include "download"
+          @response.body.must_include csv_path(id: @csv_export.id, format: 'csv')
+        end
+      end
+
+      describe "as JSON" do
+        let(:format) { :json }
+
+        it "renders json" do
+          assert_equal "application/json", @response.content_type
+          assert_response :ok
+          @csv_export.reload
+          @response.body.must_equal @csv_export.to_json
+        end
+      end
+    end
+
+    describe "a GET to show for downloaded" do
+      setup do
+        @csv_export = CsvExport.find(export.id)
```

Comment:
`let(:csv_export)` and reuse that ... then the before can be `before { csv_export.downloaded! }`